hibernate-core/src/main/java/org/hibernate/query/internal/QueryImpl.java

```diff
@@ -18,16 +19,19 @@
 public class QueryImpl<R> extends AbstractProducedQuery<R> implements Query<R> {
 	private final String queryString;
 
+	private final HQLQueryPlan queryPlan;
```

Comment:
My other concern for the above field is it might delay GC for I read somewhere it is memory intensive. Maybe a `WeakReference` is more appropriate?